### PR TITLE
Add transfer-fee function

### PR DESCRIPTION
**Pull request by barakman**

### CODE REVIEW

**Comment by barakman** [comment]
No tests with network fee > 0% as of yet.

**Comment by barakman** [comment]
> No tests with network fee > 0% as of yet.

Tests for conversions with network fee > 0% added.

Tests for liquidity functions (add, remove, get return, get cost) with network fee > 0% not yet added.

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol [hunk 1]
```diff
@@ -369,6 +371,48 @@ contract StandardPoolConverter is
...
     }
 
+    /**
+     * @dev returns the balance fee of a given reserve token
+     *
+     * @param _reserveToken    reserve token contract address
+     *
+     * @return the balance fee of the given reserve token
+     */
+    function reserveBalanceFee(IERC20 _reserveToken) public view returns (uint256) {
```
Comment (yudilevi) [inline solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:381]: I don't see why this should be public

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol [hunk 2]
```diff
@@ -369,6 +371,48 @@ contract StandardPoolConverter is
...
+        require(reserveId != 0, "ERR_INVALID_RESERVE");
+        return reserveBalanceFee(reserveId);
+    }
+
+    /**
+     * @dev returns the balance fees of both reserve tokens
+     *
+     * @return the balance fees of both reserve tokens
+     */
+    function reserveBalanceFees() public view returns (uint256, uint256) {
```
Comment (yudilevi) [inline solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:392]: I don't see why this should be public

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol [hunk 3]
```diff
@@ -369,6 +371,48 @@ contract StandardPoolConverter is
...
+        return reserveBalanceFees(1, 2);
+    }
+
+    /**
+     * @dev transfers a portion of the accumulated fees
+     */
+    function transferFees() public {
```
Comment (yudilevi) [inline solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:399]: maybe just `processNetworkFees`?

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol [hunk 4]
```diff
@@ -986,6 +1080,23 @@ contract StandardPoolConverter is
         return reserveAmounts;
     }
 
+    function reserveBalancesMinusFees(IERC20[] memory _reserveTokens) internal view returns (uint256[2] memory) {
```
Comment (yudilevi) [inline solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:1083]: fees need to be calculated based on the fee calculation formula, not by subtracting the previous balance from the current balance.
consider a case in which you have previous fee balances of 100,100 and after a swap, the balances are 200,60 - what would the fee be in this case?

**Review by yudilevi — CHANGES_REQUESTED** [review]

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol (hunk 1 above)
Comment (barakman) [reply solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:381]: @yudilevi : I added this function as part of (i.e., in favor of) adding tests for the fee before/after calling `transferFee`.
It is not even used in the contract (i.e., it is currently used **only** for testing).
We can move it to the corresponding helper contract, but the question is - why not keep it?
Just like the set of functions that we have for the reserve balances, don't we want one for the reserve balance fees?

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol (hunk 2 above)
Comment (barakman) [reply solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:392]: Same as above.

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol (hunk 3 above)
Comment (barakman) [reply solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:399]: @lbeder : suggestions/alternatives?

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol (hunk 1 above)
Comment (lbeder) [reply solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:381]: If this is only used in tests, you should move it to a `TestStandardPoolConverter.sol` helper instead

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol (hunk 3 above)
Comment (lbeder) [reply solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:399]: `processNetworkFees` isn't nad

File: solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol (hunk 1 above)
Comment (barakman) [reply solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol:381]: This is true for pretty much every external function, as well as for several public functions.
The question is as I stated it above.

**Pull request closed without merging**